src/Nethermind/Nethermind.Synchronization/FastBlocks/SyncStatusList.cs

```diff
@@ -29,74 +31,59 @@ public SyncStatusList(IBlockTree blockTree, long pivotNumber, long? lowestInsert
 
         public void GetInfosForBatch(BlockInfo?[] blockInfos)
         {
-            using ArrayPoolList<(int collected, long currentNumber)> toSent = new(blockInfos.Length);
-
             int collected = 0;
             long currentNumber = LowestInsertWithoutGaps;
-            lock (_statuses)
+
+            while (collected < blockInfos.Length && currentNumber != 0)
             {
-                while (collected < blockInfos.Length && currentNumber != 0)
+                if (blockInfos[collected] is not null)
                 {
-                    if (blockInfos[collected] is not null)
-                    {
-                        collected++;
-                        continue;
-                    }
+                    collected++;
+                    continue;
+                }
 
-                    switch (_statuses[currentNumber])
-                    {
-                        case FastBlockStatus.Unknown:
-                            toSent.Add((collected, currentNumber));
-                            _statuses[currentNumber] = FastBlockStatus.Sent;
-                            collected++;
-                            break;
-                        case FastBlockStatus.Inserted:
-                            if (currentNumber == LowestInsertWithoutGaps)
+                switch (_statuses[currentNumber])
+                {
+                    case FastBlockStatus.Unknown:
+                        if (!_cache.TryGet(currentNumber, out BlockInfo blockInfo))
+                        {
+                            blockInfo = _blockTree.FindCanonicalBlockInfo(currentNumber);
+                            if (blockInfo is not null)
                             {
-                                LowestInsertWithoutGaps--;
-                                _queueSize--;
+                                _cache.Set(currentNumber, blockInfo);
                             }
-                            break;
-                        case FastBlockStatus.Sent:
-                            break;
-                        default:
-                            throw new ArgumentOutOfRangeException();
-                    }
+                        }
 
-                    currentNumber--;
+                        blockInfos[collected] = blockInfo;
+                        _statuses[currentNumber] = FastBlockStatus.Sent;
+                        collected++;
```

Comment:
MAJOR:
I think this doesn't make sense. Now it is fairly easy for you to get duplicate marks on a block.
So multiple requests can mark a block as sent. Now this can easily lead to duplicate requests and even to marking something `Inserted` back to `Unknown`, depending on their result and order.
I think it is better to keep requests mapped 1:1 to blocks at one point in time.
Maybe we can modify the list to have a `TryMarkSent` (Unknown->Sent), `TryMarkInserted' (Sent->Inserted) and `TryMarkUnknown` (Sent->Unknown)* methods that would atomically allow only correct transitions, but not sure if it is possible.

* would also rename `Unknown` to `Pending`.

Reply:
Done